tests/eas/misfit.py

```diff
@@ -0,0 +1,400 @@
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import os
+import json
+import numpy as np
+import pandas as pd
+
+from collections import OrderedDict
+from copy import deepcopy
+from unittest import SkipTest
+
+from test import LisaTest, experiment_test
+from trace import Trace
+
+from devlib.utils.misc import memoized
+
+WORKLOAD_PERIOD_MS =  16
+
+SD_ASYM_CPUCAPACITY = 0x0040
+
+class _MisfitMigrationBase(LisaTest):
+    """
+    Base class for shared functionality of misfit migration tests
+    """
+
+    test_conf = {
+        "ftrace" : {
+            "events" : [
+                "sched_switch",
+                "sched_wakeup",
+                "sched_wakeup_new",
+                "cpu_idle",
+            ],
+            "buffsize" : 100 * 1024
+        },
+        "modules": ["cgroups", "cpufreq"],
+    }
+
+    @classmethod
+    def setUpClass(cls, *args, **kwargs):
+        super(_MisfitMigrationBase, cls).runExperiments(*args, **kwargs)
+
+    @memoized
+    @staticmethod
+    def _get_cpu_capacity(test_env, cpu):
+        """
+        :returns: The capacity value of a given CPU on the target
+        """
+        if test_env.nrg_model:
+            return test_env.nrg_model.get_cpu_capacity(cpu)
+
+        return test_env.target.read_int(
+            '/sys/devices/system/cpu/cpu{}/cpu_capacity'.format(cpu))
+
+    @memoized
+    @staticmethod
+    def _get_sd_info(test_env, cpu):
+        """
+        Read the sysfs entry of a CPU's sched domains
+
+        :returns: a dict of dicts. The first level is the integer representing
+        the sched_domain index (sched_domain<key>/* entries), the second level
+        is the sched domain field (sched_domainX/<key> entry)
+        """
+        data = test_env.target.read_tree_values(
+            '/proc/sys/kernel/sched_domain/cpu{}/'.format(cpu), depth=2
+        )
+
+        return {int(key.lstrip('domain')) : values for key, values in data.iteritems()}
+
+    @memoized
+    @staticmethod
+    def _get_hz(test_env):
+        """
+        :returns: The value of CONFIG_HZ on the target
+        """
+        return int(test_env.target.config.get('CONFIG_HZ'))
+
+    @memoized
+    @classmethod
+    def _has_asym_cpucapacity(cls, test_env):
+        """
+        :returns: Whether SD_ASYM_CPUCAPACITY is set on any sd in the system
+        """
+        for cpu in xrange(test_env.platform['cpus_count']):
+            for domain, data in cls._get_sd_info(test_env, cpu).iteritems():
+                if int(data['flags']) & SD_ASYM_CPUCAPACITY == SD_ASYM_CPUCAPACITY:
+                    return True
+
+        return False
```

Comment:
Don't what to repeat myself... but... would not be nice to have something like:

```python
test_env.target.cpus_have_asym_capacity()
```

:sunglasses:

Reply:
I think those would make sense to have in a new module e.g. `asym`.

We already have `bl` but I've always had problems with it, I think mostly because you have to set up a platform and give it core names and then it will try to match core names, etc etc.

We could go with a new module that handles those things as done here, i.e. it reads stuff from the target. We could also pave the way for DynamIQ stuff, and remove all of our weird code where we try to do what devlib.bl does without using it because of the issues mentioned above.

Reply:
I vote for a different asymmetric cpu module in devlib too